api/envoy/service/status/v2/csds.proto

```diff
@@ -0,0 +1,90 @@
+syntax = "proto3";
+
+package envoy.service.status.v2;
+
+import "envoy/admin/v2alpha/config_dump.proto";
+import "envoy/api/v2/core/base.proto";
+
+import "google/api/annotations.proto";
+import "google/protobuf/struct.proto";
+
+import "envoy/annotations/resource.proto";
+
+option java_package = "io.envoyproxy.envoy.service.status.v2";
+option java_outer_classname = "CsdsProto";
+option java_multiple_files = true;
+option java_generic_services = true;
+
+// CSDS is Client Status Discovery Service. It can be used to get the status of
+// xDS-compliant client from the management server's point of view. It can also
+// be used as an interface to get the current state of the client.
+service ClientStatusService {
+  option (envoy.annotations.resource).type = "envoy.service.status.v2.ClientStatus";
+
+  rpc StreamClientStatus(stream ClientStatusRequest) returns (stream ClientStatusResponse) {
+  }
+
+  rpc FetchClientStatus(ClientStatusRequest) returns (ClientStatusResponse) {
+    option (google.api.http).post = "/v2/discovery:client_status";
+    option (google.api.http).body = "*";
+  }
+}
+
+// Status of a config.
+enum ConfigStatus {
+  // Status info is not available/unknown.
+  UNKNOWN = 0;
+
+  // Management server has sent the config to client and received ACK.
+  SYNCED = 1;
+
+  // Config is not sent.
+  NOT_SENT = 2;
+
+  // Management server has sent the config to client but hasn’t received
+  // ACK/NACK.
+  STALE = 3;
+
+  // Management server has sent the config to client but received NACK.
+  ERROR = 4;
+}
+
+// Request for client status for a list of clients identified by Node identifier.
+message ClientStatusRequest {
+  // Node identifier for a list of clients.
+  repeated api.v2.core.Node nodes = 1;
+
+  // Metadata is used to give extra information to the management server.
+  // For example, management server can choose to only send config status if
+  // there is a metadata CONFIG_STATUS_ONLY: true.
+  google.protobuf.Struct metadata = 2;
```

Comment:
Not a fan of just adding an unstructured bag here, as we then need to standardize on the key namespace. Can we just make this a set of options that live as first class entries in the API?

Reply:
The reason I put it as metadata is that I think it is hard to standardize some of the options. For my use case, I will need project number and network name. However, I am not sure those are needed when used by other control plane. WDYT?

Reply:
The reason I put it as metadata is that I think it is hard to standardize some of the options. For my use case, I will need project number and network name. However, I am not sure those are needed when used by other control plane. WDYT?

Reply:
Well, the question is how do client and server agree on this in a standard way? How do you plan on coordinating this with multiple client and server implementations?

Reply:
It's hard. Because information needed by different server impl could be different. For example, project number and network name makes sense for a control plane running in cloud, but could have no real meaning for a control plane running on-prem. Maybe, for now, we can expose project number and network name as options for cloud use cases, and later, add options that's relevant to on-prem and such? WDYT?

Reply:
How about expressing this as a `Node` metadata matcher? Would that be sufficient? Then you can express things like project ID as Node metadata and match on them when the client presents. We already have various places we match in the API, e.g. subset LB.

Reply:
Done. Thanks!